src/OpenTelemetry.Instrumentation.ConfluentKafka/OpenTelemetryConsumeResultExtensions.cs

```diff
@@ -0,0 +1,161 @@
+    public static bool TryExtractPropagationContext<TKey, TValue>(
+        this ConsumeResult<TKey, TValue> consumeResult,
+        out PropagationContext propagationContext)
+    {
+#if NETFRAMEWORK
+        if (consumeResult == null)
+        {
+            throw new ArgumentNullException(nameof(consumeResult));
+        }
+#else
+        ArgumentNullException.ThrowIfNull(consumeResult);
+#endif
+
+        try
+        {
+            propagationContext = ExtractPropagationContext(consumeResult.Message?.Headers);
+            return true;
+        }
+        catch
+        {
+            propagationContext = default;
+            return false;
+        }
+    }
+
+    /// <summary>
+    /// Consumes a message and creates <see href="https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-spans.md#span-kind">a process span</see> embracing the <see cref="ConsumeAndProcessMessageHandler{TKey,TValue}"/>.
+    /// </summary>
+    /// <param name="consumer">The <see cref="IConsumer{TKey,TValue}"/>.</param>
+    /// <param name="handler">A <see cref="ConsumeAndProcessMessageHandler{TKey,TValue}"/>.</param>
+    /// <param name="cancellationToken">An optional <see cref="CancellationToken"/>.</param>
+    /// <typeparam name="TKey">The type of key of the <see cref="ConsumeResult{TKey,TValue}"/>.</typeparam>
+    /// <typeparam name="TValue">The type of value of the <see cref="ConsumeResult{TKey,TValue}"/>.</typeparam>
+    /// <returns>A <see cref="ValueTask"/>.</returns>
+    public static async ValueTask<ConsumeResult<TKey, TValue>?> ConsumeAndProcessMessageAsync<TKey, TValue>(
+        this IConsumer<TKey, TValue> consumer,
+        ConsumeAndProcessMessageHandler<TKey, TValue> handler,
+        CancellationToken cancellationToken = default)
+    {
+#if NETFRAMEWORK
+        if (consumer == null)
+        {
+            throw new ArgumentNullException(nameof(consumer));
+        }
+#else
+        ArgumentNullException.ThrowIfNull(consumer);
+#endif
+
+        if (consumer is not InstrumentedConsumer<TKey, TValue> instrumentedConsumer)
+        {
+            throw new ArgumentException("Invalid consumer type.", nameof(consumer));
+        }
+
+#if NETFRAMEWORK
+        if (handler is null)
+        {
+            throw new ArgumentNullException(nameof(handler));
+        }
+#else
+        ArgumentNullException.ThrowIfNull(handler);
+#endif
+
+        DateTimeOffset start = DateTimeOffset.UtcNow;
+
+        var consumeResult = instrumentedConsumer.Consume(cancellationToken);
+
+        if (consumeResult?.Message == null || consumeResult.IsPartitionEOF)
+        {
+            return consumeResult;
+        }
+
+        Activity? processActivity = null;
+        if (TryExtractPropagationContext(consumeResult, out var propagationContext))
+        {
+            TKey key = consumeResult.Message.Key;
+            object? keyAsObject = key;
```

Comment:
Just curious why this `object?` conversion is being performed?

Reply:
This cast is no longer needed as i added a generic parameter to the private `StartActivity` method to which the key is passed.